### DISCUSSION TITLE
Boussinesq for compressible fluid in NS module   ---------  Moving from Google User Group

### DISCUSSION THREAD

**Discussion by ChrocheMisawa** [idaholab/moose#17600] [Q&A Modules: Navier-Stokes] [answered]
Hello all,

I would like to ask if current moose objects are able to solve a 3D transient natural convection problems for gas, such as the cooling of a hot storage cask by its surrounding air. 

The input files under NS module folder only showed examples for incompressible material.

Digging through the old threads, I see that Finite Volume methods was under development last, but it looks like it is still only valid for incompressible fluid at the current stage?

My end goal is to model the air cooling on an cylindrical cask, and later modeling a hex cooling channel. (Combined with HeatConduction Kernel)

I have created a simple shell cylinder geometry that injects air from bottom to the top. But I was not able to use the InnerSurface as the heat surface to trigger the convection process.
The mesh file I used can be downloaded from here:
          https://drive.google.com/file/d/1-gvpLc3RptmA_OiLdVZnioWOq-bksUeT/view?usp=sharing
          https://drive.google.com/file/d/1PWP--BQMNr8AsrbJJa1VJIGj2wL6o9cl/view?usp=sharing

Can anybody suggest some suitable objects that I can use (modify) to simulate an natural convection for this problem? 

Thank you,
Zeyu Chen

**Comment by GiudGiud** [comment]
Hello

We indeed only have incompressible fluid in finite volume right now. If the temperature delta is not too high you could consider using the Boussinesq approximation. There is a PR up for FV compressible flow though, so that capability will be added soon.

All the kernels for modeling a cask with FE are already in MOOSE and its modules. 

You'll need an interface kernel for convection at the surface. I am working on one now for finite volume.
For FE, you can have a look at [ConjugateHeatTransfer](https://mooseframework.inl.gov/source/interfacekernels/ConjugateHeatTransfer.html)

Best,
Guillaume

**Reply by ChrocheMisawa** [reply]
Hello,
    Thanks for you reply!
    I looked at the ConjugateHeatTransfer example, it does not look like that it can handle the movement of the fluid phase. Or should I couple some other objects to my fluid block?
    For a cylindrical geometry that has heat source in it and cooled by surrounding air, I am trying to get some results that looks like the right contour in the picture below.

 I have managed to produce the left contour by using a ConvectiveFluxFunction BC. I assume the ConjugateHeatTransfer is also only capable to reproduce the left picture? 

![image](https://user-images.githubusercontent.com/47645670/114742717-00057400-9d1a-11eb-956d-3bde13ab5cfe.png)

**Reply by GiudGiud** [reply]
Hi

To have the movement of the fluid phase you need to look at advection kernels. The convective heat transfer interface kernel (the boundary condition should also work) just gets heat from the solid to the fluid, it does not move the fluid.

MOOSE can do laminar flow, once you have fluid advection kernels you should be able to reproduce the figure on the right. 

Guillaume

**Reply by ChrocheMisawa** [reply]
Thank you!

I looked at the advection kernels for NS, it looks like they are all for incomprehensible fluids using finite volume method (IFV).
Should I use the "ConservativeAdvection" kernel instead?

**Reply by ChrocheMisawa** [reply]
To follow up my previous comment, my question is what kernel and material objects should I use for my fluid block?

It seems that the material objects for NS module are mostly for incompressible flow. I tried to use "Air" with the [CompressibleNavierStokes] module but that did give a converged result. 
The "ConservativeAdvection" normally should stabilize the solution for me, if I am correct. But it seems that the variables used in the "module" cannot be coupled with the variable requested by "ConservativeAdvection" directly. 

Is there any example input files for compressible fluid advection that I may refer? Most of the example files only uses INS or IFV materials which I believe is not valid for my case.

Thank you,
Zeyu Chen

**Reply by GiudGiud** [reply]
If you want to use compressible flow, your best bet is to use the finite volume implementation the #17538 PR
I am not familiar with the finite element version. Most work on this is in the Pronghorn MOOSE application, for which you will have to request access.More info at [https://mooseframework.inl.gov/help/inl/index.html](https://mooseframework.inl.gov/help/inl/index.html)

Guillaume

**Reply by ChrocheMisawa** [reply]
Finite volume method also sounds interesting to me. Is #17538 open to public? It looks like it is still under development.

**Reply by GiudGiud** [reply]
As soon as the PR is up, it's technically open.
It is indeed still in development. if you encounter issues it would help to report them to @lindsayad.

@lindsayad  do you mind the early user? Or could we share the FE HLLC branch?

**Reply by lindsayad** [reply]
Anyone can do whatever they want with it :smile: Yea just know that since that branch is under development things are liable to change

**Reply by ChrocheMisawa** [reply]
> Anyone can do whatever they want with it 😄 Yea just know that since that branch is under development things are liable to change

Good to know! 
May I ask how should I utilize that branch?

Would there be any example input files, since the current examples I have are all for IFV objects.

**Reply by lindsayad** [reply]
The best place to look for examples is in `moose/modules/navier_stokes/test/tests/finite_volume/cns`

**Reply by ChrocheMisawa** [reply]
My current moose version only has "ins" in that folder, do I need to submit a pull request? Where should I submit this request?

Thank you.

**Reply by lindsayad** [reply]
`cns` will only exist if you checkout the branch on the PR #17538 that @GiudGiud linked to

**Reply by GiudGiud** [reply]
https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/checking-out-pull-requests-locally

**Reply by ChrocheMisawa** [reply]
Thank you!

I have switched to this branch and have seen these example files.
Is there any description for those test examples? It's Ok if not, I can interpret the by myself.

Best,
Zeyu Chen

**Comment by lindsayad** [comment]
Not really. You can open the ‘tests’ file which should exist in most directories. In that file there should be ‘requirements’ that outline what is being tested.

**Reply by ChrocheMisawa** [reply]
Which "test" are you referring to?

For example, I am running the "hllc_sod_shocktube_2D.i" case located in:
 moose/modules/navier_stokes/test/tests/finite_volume/cns/shock_tube_2D_cavity/

but the "tests" file in the same folder only shows:
![image](https://user-images.githubusercontent.com/47645670/115044897-f7d94000-9ea3-11eb-8db1-b69b82d3f758.png)

**Comment by lindsayad** [comment, marked as answer]
Mmm yea looks like some of the earlier tests don’t have the requirements yet

**Reply by ChrocheMisawa** [reply]
That's Ok, It's good to know that some tests will have outlines for us 👍

**Reply by ChrocheMisawa** [reply]
Hello，

I was trying to create a test natural convection case using air, by modifying the input file located at
moose/modules/navier_stokes/test/tests/ins/boussinesq/boussinesq_square.i

But it looks like the buoyancy kernel and material objects are not yet included for Compressible NSFV?
Or do they have different names?

Would you please provide a simple 2D box example case if it is achievable in the current version of moose? 
For example, cold surface on left and hot surface on right. 


Thank you,
Zeyu Chen

**Reply by lindsayad** [reply]
Note that there are four tests in upstream MOOSE in `modules/navier_stokes/test/tests/finite_volume/ins/boussinesq/tests` for Rayleigh numbers ranging from 1e3-1e6 that reproduce the results shown in https://github.com/idaholab/moose/pull/16023. I have also just pushed to my `cnsfv` branch (#17538) a couple of test cases that solve Rayleigh numbers of ~1e4 and ~1e5 using a purely compressible formulation, e.g. the tests do not use Boussinesq approximation kernels. These tests are in `modules/navier_stokes/test/tests/finite_volume/cns/natural_convection/tests`. Note that these use steady solves which are necessary for the convection discretization scheme used in the `average-boussinesq.i` input file (simple second-order central differencing). If you try to convert that input to do a transient solve you will likely see oscillations. Our current first order scheme (based on Kurganov-Tadmor without higher-order interpolation to faces) is too dissipative to yield physical results. However, I will be adding slope limiting soon such that we can bump the spatial accuracy to second order in smooth regions such that we will hopefully be able to run both steady and transient natural convection problems with a pure compressible formulation soon.

Here are the FV compressible results for the velocity magnitude for the two different Rayleigh numbers I've tested:

### Ra = 1.7e4
![Screenshot from 2021-04-28 00-58-41](https://user-images.githubusercontent.com/10248304/116368271-fde0f200-a7bc-11eb-97ac-e5d61ae35c78.png)

### Ra = 1.4e5
![Screenshot from 2021-04-28 00-59-18](https://user-images.githubusercontent.com/10248304/116368290-03d6d300-a7bd-11eb-9248-e1cf5b88dfa8.png)

**Reply by ChrocheMisawa** [reply]
Thank you!
May I ask when the slope limiting will be added?(roughly)

I also found that many of the interface kernels test failed in your branch.
And I was trying to test the ConjugateHeatTransfer module, so that I may apply a heat source in solid and import heat to the liquid, but is getting: 
"ERROR: no MaterialPropertyName parameter named "prop_getter_suffix" found."


I believe it work for the original Moose branch, is this just something with your branch specifically?

**Reply by ChrocheMisawa** [reply]
To follow my question, could you please show me how you calculated the Ra number?
I was having trouble to find the numbers for some coefficients.

Thank you,
Zeyu Chen

**Reply by lindsayad** [reply]
> May I ask when the slope limiting will be added?(roughly)

I might attempt to do this within the next week

> I also found that many of the interface kernels test failed in your branch.
> And I was trying to test the ConjugateHeatTransfer module, so that I may apply a heat source in solid and import heat to the liquid, but is getting:
> "ERROR: no MaterialPropertyName parameter named "prop_getter_suffix" found."
> 
> I believe it work for the original Moose branch, is this just something with your branch specifically?

`prop_getter_suffix` is added in my branch so yes it is definitely an issue with my branch specifically. I can't say when I'll get to fixing that but obviously it will have be fixed before my PR is merged

> To follow my question, could you please show me how you calculated the Ra number?

This script shows my computation of the Rayleigh numbers:
```python
def eta(mu, rho):
    return mu / rho
def alpha(k, rho, cp):
    return k / (rho * cp)
def beta(T):
    return 1. / T
def Tavg(Th,Tc):
    return (Th + Tc) / 2.
def dT(Th,Tc):
    return Th - Tc
def Ra(rho, beta, dT, l, g, eta, alpha):
    return rho * beta * dT * l**3 * g / (eta * alpha)

rho = 1.29
g = 9.81
gamma = 1.4
R = 8.3145
weight = 29e-3 # air
cp = gamma * (R /weight) / (gamma - 1)
Th = 400
Tc = 273.15
mu = 18.23e-6
k = 25.68e-3

l = 1e-2
print(Ra(rho, beta(Tavg(Th, Tc)), dT(Th, Tc), l, g, eta(mu, rho), alpha(k, rho, cp)))
l = 2e-2
print(Ra(rho, beta(Tavg(Th, Tc)), dT(Th, Tc), l, g, eta(mu, rho), alpha(k, rho, cp)))
```

**Reply by ChrocheMisawa** [reply]
Related to the ConjugateHeatTransfer kernel, does that mean currently the heat transfer feature at the interface between solid and fluid is not yet available? Or maybe I can try using a multi-app?



Thank you,
Zeyu Chen

**Reply by GiudGiud** [reply]
For finite volume it's currently not available. I'm adding diffusion and convection in this PR #17640 though it's not quite ready yet. 
You can do this with a multiapp approach, in which case you'll have to use Neumann (function) boundary conditions based on the differences between the fluid and solid temperature at the interface.  You'll have to use average values or try to sample points on the boundaries. I have not done the latter before. 

Best,
Guillaume

**Reply by ChrocheMisawa** [reply]
Got it, thank you!

**Reply by ChrocheMisawa** [reply]
> Note that there are four tests in upstream MOOSE in `modules/navier_stokes/test/tests/finite_volume/ins/boussinesq/tests` for Rayleigh numbers ranging from 1e3-1e6 that reproduce the results shown in #16023. I have also just pushed to my `cnsfv` branch (#17538) a couple of test cases that solve Rayleigh numbers of ~1e4 and ~1e5 using a purely compressible formulation, e.g. the tests do not use Boussinesq approximation kernels. These tests are in `modules/navier_stokes/test/tests/finite_volume/cns/natural_convection/tests`. Note that these use steady solves which are necessary for the convection discretization scheme used in the `average-boussinesq.i` input file (simple second-order central differencing). If you try to convert that input to do a transient solve you will likely see oscillations. Our current first order scheme (based on Kurganov-Tadmor without higher-order interpolation to faces) is too dissipative to yield physical results. However, I will be adding slope limiting soon such that we can bump the spatial accuracy to second order in smooth regions such that we will hopefully be able to run both steady and transient natural convection problems with a pure compressible formulation soon.
> 
> Here are the FV compressible results for the velocity magnitude for the two different Rayleigh numbers I've tested:
> 
> ### Ra = 1.7e4
> ![Screenshot from 2021-04-28 00-58-41](https://user-images.githubusercontent.com/10248304/116368271-fde0f200-a7bc-11eb-97ac-e5d61ae35c78.png)
> 
> ### Ra = 1.4e5
> ![Screenshot from 2021-04-28 00-59-18](https://user-images.githubusercontent.com/10248304/116368290-03d6d300-a7bd-11eb-9248-e1cf5b88dfa8.png)

Hi Lindsay @lindsayad 
      I'm no longer able to fetch your #17538 branch, could you please check what is wrong?


Thank you,
Zeyu Chen

**Reply by lindsayad** [reply]
Could you tell me what git commands you are using? If you fetch and then `git reset --hard lindsayad/cnsfv` then things should work. I occasionally do things like `git commit --amend` and `git rebase -i --autosquash` so this could throw off things like `git pull lindsayad cnsfv`

**Reply by ChrocheMisawa** [reply]
I used "gh pr checkout 17538"

the "git reset --hard cnsfv lindsayad/cnsfv" command didn't work, maybe because I'm currently using Mr. Guillaume's branch?

![image](https://user-images.githubusercontent.com/47645670/117371429-e0cfb180-ae95-11eb-87a4-e48c4fb74700.png)

**Reply by lindsayad** [reply]
I do not have experience with the github command-line tools, if that's what you're showing there. Yes it makes total sense... this is definitely not a fast-forward. You can do this:
```
git remote add lindsayad https://github.com/lindsayad/moose
git fetch lindsayad
git reset --hard lindsayad/cnsfv
```

**Reply by ChrocheMisawa** [reply]
I think this worked, thank you!
